luci-mod-status: add WPS control for wifi info

Ansuel opened the pull request:
Add a way to trigger and get the WPS Push Button status from the webui if supported.

Not in UTOPIA but I think it should. If the system supports this feature any type of check or trigger is absent in the webui. This feature can only be manage with the hardware button. Since the button call the exact same command, we can easily simulate this.

If this gets approved, I would ask some improvement for this... 
Is the button in the badge acceptable? How can I change this to make things better?
Also I would make the check for hostapd_cli as and env variable like we do with dnsmasq...
To clean things a bit, is it necessary to have global function instead of adding them in the view struct?

Fixes: #1072

Signed-off-by: Ansuel Smith <ansuelsmth@gmail.com>

Ansuel commented:
Also I notice that adding acl.d to the wifi index doesn't work...
Exec permission works even without acl.d permission

jow- commented:
What do you mean by that? Can you provide a specific example?

Ansuel commented:
When I was adding fs.stat for hostapd_cli for the include/60_wifi.js script, adding the permission to the acl.d section `luci-mod-status-index-wifi`, the fs.stat failed with permission denied. 
Adding it to `luci-mod-status-index` section, fs.stat correctly gets the data. (same thing for `exec`)

Also about the exec permission, as you can see there is no `exec` permission in 
```
"file": [ "list", "read", "stat" ],
```

But still 
```
				"/usr/sbin/hostapd_cli -i wlan[0-9]* wps_pbc": [ "exec" ],
				"/usr/sbin/hostapd_cli -i wlan[0-9]* wps_get_status": [ "exec" ],
				"/usr/sbin/hostapd_cli -i wlan[0-9]* wps_cancel": [ "exec" ]
```
works.

jow- commented:
Likely because some other acl enables file/exec, e.g. https://github.com/openwrt/luci/blob/master/modules/luci-mod-status/root/usr/share/rpcd/acl.d/luci-mod-status.jsonL54

You can review the effective ACL using `ubus call session list` on the cli.

Ansuel commented:
@jow- thx for the command. Anyway are we interested in this feature or should we drop this?

castillofrancodamian commented:
> @jow- thx for the command. Anyway are we interested in this feature or should we drop this?

It is an interesting function for a guest Wi-Fi network and to connect easily.

Ansuel commented:
As I said I have some question about this...

1. Is the button allowed in the wifi badge? 
2. Should I add the hostapd_cli to the SystemFeature instead of calling the stat at every pool?
3. Can I clean all the global function of 60_wifi and make them part of the view Object?
4. Is the acl rule correct? (the exec command accept any wlan[0-9]* but can the wifi interface have different naming in openwrt?)

jow- commented:
1. Not sure, can you provide a screenshot to visualize what you mean?
2. Yes, I think that would make sense
3. Sure, but the function is not global, it is local to the `60_wifi.js` view class. Why do you want to move it?
4. We probably need to wrap this call in a `/usr/libexec/wps_action` utility script which takes care of enforcing the parameter. The current pattern `"/usr/sbin/hostapd_cli -i wlan[0-9]* wps_pbc"` is too broad since it would allow an arbitrary amount of parameters after `wlan...`.

The utility script could look like this:
```bash
#!/bin/sh

case "$1" in
  pbc|get_status|cancel)
    if [ -e "/sys/class/net/$2" ]; them
      exec /usr/sbin/hostapd_cli -i "$2" "wps_$1"
    else
      echo "No such interface: $2" >&2
      exit 1
    fi
  ;;
  *)
    echo "Usage: $0 {pbc,get_status,cancel} ifname" >&2
    exit 1
  ;;
esac
```
And a corresponding ACL rule:
```
"/usr/libexec/wps_action": [ "exec" ]
```

Ansuel commented:
1. Screen
![Immagine 2020-10-10 223234](https://user-images.githubusercontent.com/20289090/95664490-93ca6d80-0b48-11eb-8698-e7cffe803320.jpg)
Will address the other suggestion thanks.

jow- commented:
Hmmm, looks a bit out of place. I think it would look better if you omit the label, so that it looks like this:

![grafik](https://user-images.githubusercontent.com/2528802/95664729-ddb45300-0b4a-11eb-99e5-716a34d7dc4a.png)

Ansuel commented:
The best thing would be have a small button aligned to the size of the text but i think it's not available in luci css.

dangowrt commented:
We actually got a ubus interface for this, so instead you could `ubus call hostapd.wlan0 wps_start` or, even better, use `rpcd` for that.

Ansuel commented:
@dangowrt just as an info... where this was documented? Don't know if it's me but it's hard to find what ubus expose :(

Anyway this means that wps can be supported even without hostapd-utils using ubus... Can you confirm this?

dangowrt commented:
> @dangowrt just as an info... where this was documented? Don't know if it's me but it's hard to find what ubus expose :(
Use `ubus -v list` to explore this kinda a self-documenting API ;)

> 
> Anyway this means that wps can be supported even without hostapd-utils using ubus... Can you confirm this?
Yes. And it works for wpa_supplicant just the same way.

Ansuel commented:
Didn't know -v list could be used also with params (`hostapd.wlan0`) nice to know

Well I have to make a patch to add support for get status ahahah 
This needs to wait :(

dangowrt commented:
> Also no wps_get_status :(

I could add this easily if needed. Is it?

Ansuel commented:
@dangowrt if you want and can implement this quickly it would be very good...
```
root@No-Lag-Router:~# hostapd_cli -i wlan0 wps_get_status
PBC Status: Timed-out
Last WPS result: None
```

Best thing would be convert this to a json output

Ansuel commented:
@dangowrt I implemented a test version anyway...

dangowrt commented:
https://git.openwrt.org/?p=openwrt/staging/dangole.git;a=commit;h=4c3d9f113d86ecfebbd4c8ef0531be45e5ed3a3b

Ansuel commented:
Thx a lot. Is CONFIG_WPS enabled by default?

dangowrt commented:
> Thx a lot. Is CONFIG_WPS enabled by default?

Currently we got `CONFIG_WPS` enabled only for the `-full` and `-p2p` variants, but not in the `-basic` and `-mini` variants which are installed by default.

Ansuel commented:
@dangowrt sorry if I bother you again... Since CONFIG_WPS can be disabled... Can we make a patch to add wps to the `has_feature`
function?  (src/utils/build_features.h)

Ansuel commented:
@jow- I updated this... this depends on 2 patch
1. https://git.openwrt.org/?p=openwrt/staging/dangole.git;a=commit;h=4c3d9f113d86ecfebbd4c8ef0531be45e5ed3a3b
2. The `CONFIG_WPS` in`hostapd -v` that still needs to be created

Since now wps is supported without hostapd_cli, I fixed the wireless page to show the pushbutton option even without hostapd_cli.

dangowrt commented:
> @dangowrt sorry if I bother you again... Since CONFIG_WPS can be disabled... Can we make a patch to add wps to the `has_feature`
> function? (src/utils/build_features.h)
> Can't find any other way to check this....

Can't you just check if the ubus methods are present? That'd be the most obvious way to do it imho.

dangowrt commented on modules/luci-mod-status/htdocs/luci-static/resources/view/status/include/60_wifi.js:
```diff
@@ -2,90 +2,45 @@
 
+	WPSTranslateTbl: {
+		WPS_PBC_STATUS_DISABLE: _('Disabled'),
+		WPS_PBC_STATUS_ACTIVE: _('Active'),
```
Probably it'll be easier then I'd just output those shorted strings right away then...

Ansuel replied:
I actually like your implementation that prints the symbol... That way we can enum all the different strings and can be translated by the webui... (in short I had to do the same exact thing even if ubus reported the text directly)

Ansuel commented:
Currently luci uci the presence of hostapd_cli to check if wps is supported. I think that calling ubus to check if there are wps entries in any wlan interface is not good... (imho check if a command doesn't fail to make sure a feature is supported is wrong...)
Also luci should call hostapd.(random present interface), so every time it should load all the wifi interface, take one and call ubus (and check if the command fail)

IMHO it's much cleaner to add support to the -v command as it's already done to check if hostapd supports 11ac 11r and other feature

dangowrt commented:
@Ansuel I've updated the status API to make things easier, see my staging tree:
https://git.openwrt.org/?p=openwrt/staging/dangole.git;a=commit;h=b51e22fa2b75d0c05dfbb02d0f120cbccbd65395

Ansuel commented:
Anyway the change would be mainly this in `src/utils/build_features.h`
```
#ifdef CONFIG_WPS
	if (!strcmp(feat, "wps"))
		return 1;
#endif
```

dangowrt commented:
> Anyway the change would be mainly this in `src/utils/build_features.h`
> 
> ```
> #ifdef CONFIG_WPS
> 	if (!strcmp(feat, "wps"))
> 		return 1;
> #endif
> ```

Updated in https://git.openwrt.org/?p=openwrt/staging/dangole.git;a=commit;h=f1704f2fc8c03c544ba3b314121e7a848457979a

Ansuel commented:
Again thx a log for the patch. Will update this while I wait for a review from jow

Ansuel commented:
The patch has been merged and this is ready.

jow- commented on modules/luci-mod-status/root/usr/share/rpcd/acl.d/luci-mod-status.json:
```diff
@@ -120,12 +120,12 @@
 		"read": {
 			"ubus": {
-				"iwinfo": [ "assoclist" ]
+				"iwinfo": [ "assoclist", "wps_get_status" ]
```
Shouldn't that be `"hostapd.*": [ "wps_status" ]` ?

Ansuel replied:
Yes you are right.

jow- commented:
Merged, thanks!
